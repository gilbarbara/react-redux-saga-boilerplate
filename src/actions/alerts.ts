import { ReactNode } from 'react';
import { uuid } from '@gilbarbara/helpers';
import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from '~/modules/helpers';

import { ActionTypes } from '~/literals';

import { ShowAlertOptions } from '~/types';

export const hideAlert = createAction<string>(ActionTypes.ALERTS_HIDE);

export const showAlert = createAction(
  ActionTypes.ALERTS_SHOW,
  (message: ReactNode, options: ShowAlertOptions) => {
    const timeout = options.type === 'error' ? 0 : 5;

    return actionPayload({
      id: options.id || uuid(),
      icon: options.icon || 'info-o',
      message,
      position: options.position || 'bottom-right',
      type: options.type || 'neutral',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    });
  },
);
