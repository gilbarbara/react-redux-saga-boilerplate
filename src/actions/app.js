// @flow
/**
 * @module Actions/App
 * @desc App Actions
 */

import uuid from 'uuid/v4';
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export { goBack, go, push, replace } from 'modules/history';

export const { hideAlert, showAlert, switchMenu } = createActions({
  [ActionTypes.SWITCH_MENU]: (query: string) => ({ query }),
  [ActionTypes.HIDE_ALERT]: (id: string) => ({ id }),
  [ActionTypes.SHOW_ALERT]: (message: string, options: Object) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || uuid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
});
