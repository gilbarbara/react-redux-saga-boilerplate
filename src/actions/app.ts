/**
 * @module Actions/App
 * @desc App Actions
 */
import React from 'react';
import { nanoid } from 'nanoid';

import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { ShowAlertOptions } from 'types';

export { go, goBack, push, replace } from 'modules/history';

export const hideAlert = createAction(ActionTypes.HIDE_ALERT, (id: string) => ({ id }));

export const showAlert = createAction(
  ActionTypes.SHOW_ALERT,
  (message: React.ReactNode, options: ShowAlertOptions) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || nanoid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
);
