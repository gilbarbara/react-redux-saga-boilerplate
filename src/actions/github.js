// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { githubGetRepos: getRepos } = createActions({
  [ActionTypes.GITHUB_GET_REPOS]: (query: string) => ({ query }),
});
