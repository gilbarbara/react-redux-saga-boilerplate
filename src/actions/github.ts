/**
 * @module Actions/User
 * @desc User Actions
 */
import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

export const getRepos = createAction(ActionTypes.GITHUB_GET_REPOS_REQUEST, (query: string) => ({
  query,
}));
