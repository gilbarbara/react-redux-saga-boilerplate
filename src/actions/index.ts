export * from './app';
export * from './github';
export * from './user';
