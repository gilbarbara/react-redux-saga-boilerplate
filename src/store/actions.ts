export { alertHide, alertShow } from './slices/alerts';
export { setAppOptions } from './slices/app';
export { getRepos, getReposFailure, getReposSuccess } from './slices/github';
export { login, loginSuccess, logOut, logOutSuccess } from './slices/user';
