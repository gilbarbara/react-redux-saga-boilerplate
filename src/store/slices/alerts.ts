import { ReactNode } from 'react';
import { uuid } from '@gilbarbara/helpers';
import type { PayloadAction } from '@reduxjs/toolkit';
import { createSlice } from '@reduxjs/toolkit';

import { actionBody } from '~/modules/helpers';

import { AlertData, AlertsState, ShowAlertOptions } from '~/types';

export const alertsState: AlertsState = {
  data: [],
};

export const alertsSlice = createSlice({
  name: 'alerts',
  initialState: alertsState,
  reducers: {
    alertHide: (draft, { payload }: PayloadAction<string>) => {
      draft.data = draft.data.filter(d => d.id !== payload);
    },
    alertShow: {
      reducer: (draft, { payload }: PayloadAction<AlertData>) => {
        draft.data.push(payload);
      },
      prepare: (message: ReactNode, options: ShowAlertOptions) => {
        const timeout = options.type === 'error' ? 0 : 5;

        return actionBody({
          id: options.id ?? uuid(),
          icon: options.icon ?? 'info-o',
          message,
          position: options.position ?? 'bottom-right',
          type: options.type ?? 'neutral',
          timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
        });
      },
    },
  },
});

export const { alertHide, alertShow } = alertsSlice.actions;

export default alertsSlice.reducer;
