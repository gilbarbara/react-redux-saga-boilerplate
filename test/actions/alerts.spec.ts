import { hideAlert, showAlert } from '~/actions/alerts';

describe('actions/alerts', () => {
  describe('showAlert', () => {
    it('should return an action with type `error`', () => {
      expect(showAlert('Alright!', { id: 'test', type: 'error' })).toMatchSnapshot();
    });

    it('should return an action with type `success`', () => {
      expect(showAlert('Alright!', { id: 'test', type: 'success', timeout: 10 })).toMatchSnapshot();
    });
  });

  describe('hideAlert', () => {
    it('should return an action', () => {
      expect(hideAlert('test')).toMatchSnapshot();
    });
  });
});
