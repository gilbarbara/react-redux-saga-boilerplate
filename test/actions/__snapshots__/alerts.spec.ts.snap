// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`actions/alerts > hideAlert > should return an action 1`] = `
{
  "payload": "test",
  "type": "ALERTS_HIDE",
}
`;

exports[`actions/alerts > showAlert > should return an action with type \`error\` 1`] = `
{
  "meta": undefined,
  "payload": {
    "icon": "info-o",
    "id": "test",
    "message": "Alright!",
    "position": "bottom-right",
    "timeout": 0,
    "type": "error",
  },
  "type": "ALERTS_SHOW",
}
`;

exports[`actions/alerts > showAlert > should return an action with type \`success\` 1`] = `
{
  "meta": undefined,
  "payload": {
    "icon": "info-o",
    "id": "test",
    "message": "Alright!",
    "position": "bottom-right",
    "timeout": 10,
    "type": "success",
  },
  "type": "ALERTS_SHOW",
}
`;
